crates/federate/src/worker.rs

```diff
@@ -166,203 +235,534 @@ impl InstanceWorker {
+  #[serial]
+  async fn test_stats(data: &mut Data) -> LemmyResult<()> {
+    tracing::debug!("hello world");
+
+    // first receive at startup
+    let rcv = data.stats_receiver.recv().await.unwrap();
+    tracing::debug!("received first stats");
+    assert_eq!(data.instance.id, rcv.state.instance_id);
+    // assert_eq!(Some(ActivityId(0)), rcv.state.last_successful_id);
+    // let last_id_before = rcv.state.last_successful_id.unwrap();
+
+    let sent = send_activity(data.person.actor_id.clone(), &data.context, true).await?;
+    tracing::debug!("sent activity");
+    // receive for successfully sent activity
+    let inbox_rcv = data.inbox_receiver.recv().await.unwrap();
+    let parsed_activity = serde_json::from_str::<WithContext<Value>>(&inbox_rcv)?;
+    assert_eq!(&sent.data, parsed_activity.inner());
+    tracing::debug!("received activity");
+
+    let rcv = data.stats_receiver.recv().await.unwrap();
+    assert_eq!(data.instance.id, rcv.state.instance_id);
+    assert_eq!(Some(sent.id), rcv.state.last_successful_id);
+    tracing::debug!("received second stats");
+
+    data.cleanup().await?;
+
+    // it also sends state on shutdown
+    let rcv = data.stats_receiver.try_recv();
+    assert!(rcv.is_ok());
+
+    // nothing further received
+    let rcv = data.stats_receiver.try_recv();
+    assert_eq!(Some(TryRecvError::Disconnected), rcv.err());
+    let inbox_rcv = data.inbox_receiver.try_recv();
+    assert_eq!(Some(TryRecvError::Disconnected), inbox_rcv.err());
+
+    Ok(())
+  }
+
+  #[test_context(Data)]
+  #[tokio::test]
+  #[traced_test]
+  #[serial]
+  async fn test_send_100(data: &mut Data) -> LemmyResult<()> {
+    tracing::debug!("hello world");
+
+    // first receive at startup
+    let rcv = data.stats_receiver.recv().await.unwrap();
+    tracing::debug!("received first stats");
+    assert_eq!(data.instance.id, rcv.state.instance_id);
+    // assert_eq!(Some(ActivityId(0)), rcv.state.last_successful_id);
+    // let last_id_before = rcv.state.last_successful_id.unwrap();
+    let mut sent = Vec::new();
+    for _ in 0..100 {
+      sent.push(send_activity(data.person.actor_id.clone(), &data.context, false).await?);
     }
-    if (Utc::now() - self.last_incremental_communities_fetch) > *FOLLOW_ADDITIONS_RECHECK_DELAY {
-      // process additions every minute
-      let (news, time) = self
-        .get_communities(self.instance.id, self.last_incremental_communities_fetch)
-        .await?;
-      self.followed_communities.extend(news);
-      self.last_incremental_communities_fetch = time;
+    sleep(2 * *WORK_FINISHED_RECHECK_DELAY).await;
+    tracing::debug!("sent activity");
+    compare_sent_with_receive(data, sent).await?;
+
+    Ok(())
+  }
+
+  #[test_context(Data)]
+  #[tokio::test]
+  #[traced_test]
+  #[serial]
+  /// this test sends 15 activities, waits and checks they have all been received, then sends 50,
+  /// etc
+  async fn test_send_15_50_200(data: &mut Data) -> LemmyResult<()> {
```

Comment:
You can decrease the number of activities sent here and above to speed up testing, dont think theres any need for 100+

Reply:
Since for the test the parallel count is set to 10, I wanted every group to have at least more than one chunk of simultaneous sends. I'll reduce it to send_40 and send_15_20_30